### PR TITLE
API proposal for notebook extensibility

### PR DESCRIPTION
**Pull request by rchiodo**
Draft proposal for extended extensibility for notebooks. For this item here:
https://github.com/microsoft/vscode-python/issues/10505

Bunch of stuff
1) Make cell outputs have extra data for correlation between execution and render providers. 
- **Use case**: Execution puts a special tag for the file name for the interactive window. This indicates to an NotebookOutputTransformer to add links into the error output.
2) Add execution information to a document and a cell (this would be the kernel to use). 
- **Use case**: Allows a kernel to be started with the correct information. Language isn't enough.
3) Split execution from NotebookDocument creation (at least I think the NotebookProducer translates ipynb files to NotebookDocument objects). 
- **Use case**: One extension translates documents, while multiple might provide just execution for ipynb files.
4) Make execution fire events as there are multiple stages of output. 
- **Use case**: Streaming data as it runs. Clearing data part way through.
5) Create a NotebookOutputTransformer extensibility point. 
- **Use case**: Modifying error to have line numbers and clicks. Modifying images to allow for opening in a plot viewer.
6) Create a NotebookExecutionObserver extensibility point.
- **Use case**: Variable explorer, gather. Allows for a separate kernel to be kept in sync.

### CODE REVIEW

File: src/vs/vscode.proposed.d.ts [hunk 1]
```diff
@@ -1524,15 +1525,20 @@ declare module 'vscode' {
...
+	}
 
 	export interface NotebookCell {
 		readonly uri: Uri;
 		handle: number;
 		language: string;
 		cellKind: CellKind;
 		outputs: CellOutput[];
-		getContent(): string;
+		executionInfo: NotebookExecutionInfo;
```
Comment (DonJayamanne) [inline src/vs/vscode.proposed.d.ts]: How about we call this `data`.
Something that gets passed around, akin to `metadata`.
How it is used is upto the extension authors.

File: src/vs/vscode.proposed.d.ts [hunk 2]
```diff
@@ -1541,6 +1547,7 @@ declare module 'vscode' {
 		readonly isDirty: boolean;
 		languages: string[];
 		cells: NotebookCell[];
+		rootExecutionInfo: NotebookExecutionInfo;
```
Comment (DonJayamanne) [inline src/vs/vscode.proposed.d.ts]: Same here, `data`, i.e. state bag to dump anything in there.

File: src/vs/vscode.proposed.d.ts (hunk 1 above)
Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: Good idea

File: src/vs/vscode.proposed.d.ts [hunk 3]
```diff
@@ -1554,24 +1561,72 @@ declare module 'vscode' {
...
+
 	export interface NotebookOutputRenderer {
 		/**
 		 *
 		 * @returns HTML fragment. We can probably return `CellOutput` instead of string ?
 		 *
 		 */
 		render(document: NotebookDocument, cell: NotebookCell, output: CellOutput, mimeType: string): string;
-		preloads?: Uri[];
+		preloads?: Uri[]; // rchiodo: What is this for?
```
Comment (DonJayamanne) [inline src/vs/vscode.proposed.d.ts:1622]: To preload `react` or other libraries that may be required when rendering the output.

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts:1622]: Ah so they woudln't have to be in the returned string

File: src/vs/vscode.proposed.d.ts [hunk 4]
```diff
@@ -1554,24 +1561,72 @@ declare module 'vscode' {
...
+	export interface NotebookExecution {
+		readonly executionCount: number;
+		executeCell(document: NotebookDocument, cell: NotebookCell): NotebookCellExecutionResult;
+	}
+
+	/**
+	 * Allows observing the execution of cells from a NotebookExecution
+	 */
+	export interface NotebookExecutionObserver {
+		onExecuteCell(document: NotebookDocument, cell: NotebookCell, eventProvider: NotebookCellExecutionResult): void;
```
Comment (DonJayamanne) [inline src/vs/vscode.proposed.d.ts]: Not sure this is an observer, in fact it provides yet another hook to add event handlers into `eventProvider`. 

I.e. there are two or more steps:
1. registerNotebookExecutionObserver
2. eventProviders.onClear....

I would have expected the observer to have methods on it to get the information...
as follows:
```typescript
	export interface NotebookCellExecutionObserver {
		onCellOutput(document: NotebookDocument, cell: NotebookCell...)
        onCel.....
	}

Personally I like this, then there's very little API that needs to be exposed from VSCode.
Else they need to expose a lot about the structure of how things are managed and events exposed, and the like.
```

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: Okay but it would need onExecute as well. You're just eliminating the extra code of signing up for the events.

File: src/vs/vscode.proposed.d.ts (hunk 3 above)
Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts:1622]: Yes. this way the preload is a one time task and loading of resources (first time) is delegated to VS Code.

File: src/vs/vscode.proposed.d.ts (hunk 4 above)
Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: Not really (though that's a benefit), and it allows VSC to  change their API without having to worry much about the consumders
Else with the other path VSC (API publisher) will be forced to expose a specific structure that they need to adhere to..
E.g. its possible we could have separate classes that deal with execution, and another class for deleting cells (hence clear event is somewhere else), etc... 
This may not be the case, but I think it allows the publisher to evolve/change without having to worry about breaking any API...

Basically if I was exposing an API, I'd allow for my (publisher) end to have most flexibility..

File: src/vs/vscode.proposed.d.ts [hunk 5]
```diff
@@ -1580,7 +1636,23 @@ declare module 'vscode' {
 			provider: NotebookProvider
 		): Disposable;
 
-		export function registerNotebookOutputRenderer(type: string, outputSelector: NotebookOutputSelector, renderer: NotebookOutputRenderer): Disposable;
+		export function registerNotebookExecution(
+			execution: NotebookExecution
```
Comment (DonJayamanne) [inline src/vs/vscode.proposed.d.ts:1632]: Looking at this API, who is responsible for executing cells?
VSC or extension author?
If its extension author then we need a way to provide results to VSC in response to an execute request or similar.

File: src/vs/vscode.proposed.d.ts [hunk 6]
```diff
@@ -1524,15 +1525,17 @@ declare module 'vscode' {
...
+	export type CellOutput = CellStreamOutput | CellErrorOutput | CellDisplayOutput & { metadata: { [key: string]: any} }; // Metadata can be put here by an execution so that a renderer can use it
 
 	export interface NotebookCell {
 		readonly uri: Uri;
 		handle: number;
 		language: string;
 		cellKind: CellKind;
 		outputs: CellOutput[];
-		getContent(): string;
+		metadata: { [key: string]: any }
```
Comment (DonJayamanne) [inline src/vs/vscode.proposed.d.ts]: Looking at this structure, the `NotebookCell` is mutable.

* Should we instead have a separate way of adding metadata to cells?
* Cannot the extension authors add metadata in their own extension and have it keyed to the notebookCell? This way VSC is free of storing metadata?
* If the metadata is unstructured, its not really useful to other extension authors, hence see little benefit here

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: Some metadata is valid, such as:
* Readonly
* Selected

Though, they could be upgraded to standard properties on the cell..So other extensions can utilize them...

File: src/vs/vscode.proposed.d.ts (hunk 5 above)
Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts:1632]: VSC is responsible. THat's what the NotebookExecutionResult is for. It has events they have to listen to for output etc.

File: src/vs/vscode.proposed.d.ts (hunk 6 above)
Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: Yeah but then one extension cannot communicate metadata to another cell. Or in the case we have now, one product to another.

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: > we have now, one product to another.

Hmm, doesn't feel right. Its basically `any`, hence to me `no defined api`.
But do we have a use case of where we'd like one extension to expose metadata to another.

I can only think of gather, is that the usecase.

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: User puts metadata in the jupyter notebook editor. We have a custom renderer that reads it and uses it for output. 

It's also where the kernel spec information would go so we'd know what kernel to start. Otherwise we're predefining what a kernel is and that's a jupyter specific thing.

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: Not sure we need the concept of execution.

Why not?:
* Extensions provide results and tells VSC what to render
* VSC then fires events saying `onResult` or `onRendered` or similar
    * Now other extensions can listen in on them and know something has happened...
    * Similarly VSC can fire onEdited events...
* Extensions provide results for document wide results as `Document Content`. VSC Doesn't need to know whether its tied to a cell or not, all VSC needs to know is what needs to be rendered, just as with Cell results. 

It just deals with rendering something. Extensions execute ells and provide output. The missing bit is hooks for buttons, those handlers are what cause extension authors to provide results..

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: It has to be 'any' so it can not be bound to the jupyter specific implementation.

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: > It's also where the kernel spec information would go so we'd know what kernel to start.

VSC doesn't need this. Extension authors can get this from the file or notebook, we do'nt need VSC API to provide this. I.e. only extensions need this information.

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: Execution providers don't have access to the notebook (unless they read it from disk?) And how would a change in the kernel be propagated around?

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: The notebookProvider translates the json into the NotebookDocument format. An execution provider can then execute the notebook. Hmm, maybe metadata doesn't work here either, but rather a more concrete ExecutionInfo or something along those lines.

However the other metadata in a notebook or cell that updates in memory, how would that be communicated between extensions?

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: > And how would a change in the kernel be propagated around?

Why do we need to propagate this to other extensions?
I'm assuming VSC will provide access to the notebook data (raw string) or similar (or Uri).

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: > how would that be communicated between extensions?

Not sure why extensions need them? I'd like to understand a usecase.
As suggested earlier, why can't extensions provide metadata along with results when calling VSC to render results.
Then VSC can propagate this to other extensions saying `here you go I received these results+metadata, do what you want with it`

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: Extension 1: NotebookProvider = turns .foobar files into Notebooks
Extension 2: Renders results and looks at the 'collapsed/expanded' metadata that might be on a cell.
Extension 3: Executes notebooks. Knows nothing of metadata, just passes it along.

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: I agree I don't like the metadata: any thing either. Trying to think of some other way to communicate information in the Notebook (both in memory and from disk) to other extensions

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: Kernel spec information might be specific enough we create a VS code euphemism for it. I think I'll try that. But other metadata is weirder.

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: > 'collapsed/expanded' 

These are some of the generic items that will have to be exposed by VSC. I.e. it is UI state, hence VSC will have to expose them to extension authors vs extension authors putting some randomnly named metadata to indicate a cell is expanded or collapsed, After all that initial information needs to come from VSC.
I.e. `collpased/expanded/selected/focused/readonly` and others would be VSC specific and I'd expect to have specific properties on cells for these.

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: > communicate information in the Notebook (both in memory and from disk) to other extensions

Still feel results would provide enough information that can be shared with other extensions.

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: You mean the metadata tag I added to every output?

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: That might be enough. It means a NotebookProvider can't communicate any data in the base notebook though.

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: > You mean the metadata tag I added to every output?

That's great.
I think we need a way for VSC to broadcast this information to other extensions.

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: They would all get it. NotebookExecutionObserver would get it on its different callbacks. The NotebookOutputTransformer would get it when transforming and the NotebookOutputRenderer would see it when rendering. 

I'm still worried about all the extra information at the bottom of an ipynb. What if you need that to start your kernel? Or your language server? I guess you could read it from disk too though.

File: src/vs/vscode.proposed.d.ts [hunk 7]
```diff
@@ -1554,24 +1564,76 @@ declare module 'vscode' {
...
+
+	/**
+	 * Allows observing the execution of cells from a NotebookExecution
+	 */
+	export interface NotebookExecutionObserver {
+		onExecuteStart(document: NotebookDocument, cell: NotebookCell): void;
+		onCellOutput(document: NotebookDocument, cell: NotebookCell, outputs: CellOutput[]): void;
+		onCellClear(document: NotebookDocument, cell: NotebookCell, wait: boolean): void;
+		onDocumentOutput(document: NotebookDocument, cell: NotebookCell, outputs: CellOutput[]): void;
+		onExecuteComplete(document: NotebookDocument, cell: NotebookCell, executionCount: number): void;
```
Comment (DonJayamanne) [inline src/vs/vscode.proposed.d.ts:1607]: Not sure about this, now that we have `metadata` on each cell, why do we need these?

Ie. why would one need `onExecutionStart` and `onExecuteionComplete`.
Does an extension need to know something has started? I would have expected an extension to merely know execution completed, and that information cab be derived from the fact that something was rendered for a cell.

I'm thinking of `gather`, we only need to know when something has completed.
Similarly `cellClear` that can be communicated with `onCellOutput` having empty `outputs`.

File: src/vs/vscode.proposed.d.ts [hunk 8]
```diff
@@ -1554,24 +1564,76 @@ declare module 'vscode' {
...
+	/**
+	 * Allows transformation of output from raw execution. Called during execution.
+	 */
+	export interface NotebookOutputTransformer {
+		/**
+		 *
+		 * @returns the next item in the chain
+		 *
+		 */
+		transformOutput(document: NotebookDocument, cell: NotebookCell, output: CellOutput): CellOutput;
```
Comment (DonJayamanne) [inline src/vs/vscode.proposed.d.ts]: Isn't the the responsibility of a rendered?
I.e. extension authors can do their transformation and return the result in `NotebookOutputRender`.
Making it unnecessary for VSC to ask for extensions to transform output, when VSC is only ever concerned with the `string output` for rendering.

File: src/vs/vscode.proposed.d.ts (hunk 7 above)
Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts:1607]: OnCellClear is a result of the jupyter message - clear a cell. You need to have this in order to know when to clear (might be now, might be after). 

The Observer doesn't have the underlying document, it just has the events. There'd be no onCellOutput for actually clearing the output. 

onExecutionStart is necessary so you can kick off your own execution (like the variable explorer would need)

File: src/vs/vscode.proposed.d.ts (hunk 6 above)
Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: > I guess you could read it from disk too though.

I would except VSC to provide raw access to the file or provide a Uri. At the end of the day `ipynb` is one format that backs notbooks. But it doesn't have to be an `ipynb` file.

File: src/vs/vscode.proposed.d.ts (hunk 8 above)
Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: No this transforms one output type to another. It doesn't render anything.

File: src/vs/vscode.proposed.d.ts (hunk 6 above)
Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: I'm not saying we throw away ipynb, just saying that VSC will provide some way to access the information, at a minimum the Uri (even if its remote we should be able to get the data when running remotely).

File: src/vs/vscode.proposed.d.ts (hunk 8 above)
Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: Say given a image, it turns it into an ipywidget request (although not sure how that's going to work).

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: Yeah how would ipywigets work at all with this model. They need the comm with the kernel. I guess we'd have to put them inside the same extension that implements execution

File: src/vs/vscode.proposed.d.ts (hunk 7 above)
Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts:1607]: > OnCellClear is a result of the jupyter message - clear a cell. You need to have this in order to know when to clear (might be now, might be after).

As mentioned earlier, this can be achieved by telling VSC to render and empty result. Empty result = clear output.

> onExecutionStart is necessary so you can kick off your own execution (like the variable explorer would need)

Hmm, I'd expect this to only come up `after` execution, not while.
I.e. cell has been executed, now lets display the variable explorer.

File: src/vs/vscode.proposed.d.ts (hunk 8 above)
Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: The use case for the transform is actually for our error line modification. It transforms an error into something with links in it.

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: But you're right we could just have a text/error handler.

File: src/vs/vscode.proposed.d.ts (hunk 7 above)
Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts:1607]: You don't tell VSC to render. I think you're missing the mental model here.

There's a NotebookDocument. It has cells in it. VS code owns this object. When a user clicks 'run cell', VS code finds the NotebookExecution and calls executeCell on it. It then signs up for the events in the NotebookExecutionResult. 

onClear comes in as an event. VS code must then clear the cell outputs in its representation of the Document. There are no extensions that change the Document's outputs. VS code does that.

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts:1607]: As far as the variable explorer, it needs to execute the same code as the real kernel so that it's internal state is the same. That's why it needs to know when the cell begins.

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts:1607]: Why not do this:

* VSC fires events when buttons are clicked (fire commands) as they do today
* Extensions do their stuff and send results to VSC

The only thing thats missing is `trigger for extension to start executing`. I.e. if we looked at the existing model (what VSC have today), VSC does not have a concept of execution it only renders stuff. It handles buttons clicks like `ADD CELL`,etc..

I think we can still do this by just letting VSC expose events for `executeCell` or similar, I.e.g `Command` with a cell.
This way we can have commands for `Clear Cell`, `Run Cell`, etc.. 

If feel that fits with little change to existing API, and gives us what we (extension authors) need.

File: src/vs/vscode.proposed.d.ts (hunk 8 above)
Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: > Say given a image, it turns it into an ipywidget request (although not sure how that's going to work).

Again, the only purpose of this is to render something, Hence this is satisfied with what we have today.

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: > They need the comm with the kernel. I guess we'd have to put them inside the same extension that implements execution

We use `preloads` and then render the result. The render result contains `javascript` that has the necessary context (loaded from preload), that can then communicate back, to extension with `onMessage` or some other mechanism. Cuz VSC cannot use `onCellExecuted` at all, cuz IPYWidgets needs to be able to send/receive a stream and current VSC API wpont' work and i think its overkill to use that interface to transmit ipywidget data back and forth.

File: src/vs/vscode.proposed.d.ts (hunk 7 above)
Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts:1607]: I think you're over simplifying. The Command for 'execute' cell is the same as the API here. And it doesn't allow other extensions to listen to the command. Commands are single listener at a time.

And how does this part work?
> Extensions do their stuff and send results to VSC

How do they send results? The results aren't a thenable action. They're a multilevel set of messages that come back. Hence the NotebookExecutionResult. I'd much rather use an observable, but they don't support that in VS code.

File: src/vs/vscode.proposed.d.ts (hunk 8 above)
Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts]: > It transforms an error into something with links in it.

So why don't we have our own result that renders this.
I wouldn't want other extensions coming and meddling with the `<html>` result that I generated for display in VSC.

That could result in unexpected/unwanted results. I.e. one extension modifying the Html provided by another extension. Its basically `find and replace HTML`, not really extensible to me.

File: src/vs/vscode.proposed.d.ts (hunk 7 above)
Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts:1607]: And I think the clear I'm talking about is not the same clear you're thinking of. My clear comes from this:

```
from IPython.display import clear_output

for i in range(10):
    clear_output()
    print("Hello World!")
```

Not a button on a cell. It's stuck in the middle of outputting.

Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts:1607]: > And it doesn't allow other extensions to listen to the command. Commands are single listener at a time.

Again, the only piece missing is triggering extension to execute cells. Should we first try to solve that trhan adding concept of execution and the like. 
I.e. solve problem of VSC sending UI events to extension.

File: src/vs/vscode.proposed.d.ts (hunk 8 above)
Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts]: Yeah maybe, I want to leave it there to see what Kai thinks. The transform pipeline was his idea.

File: src/vs/vscode.proposed.d.ts (hunk 7 above)
Comment (DonJayamanne) [reply src/vs/vscode.proposed.d.ts:1607]: > Hence the NotebookExecutionResult. I'd much rather use an observable, but they don't support that in VS code.

Why does VSC need observable. Its extension that needs observable to update the ui.
In react we don't have observable, we merely render the data as and when required. I.e. observable is merely used to `react` to sometning that happened. In this case, the result of the `reaction` is to render a result.

Comment (rchiodo) [reply src/vs/vscode.proposed.d.ts:1607]: The execution needs the observable because it's a mulitstep thing. Rendering doesn't.

I don't think discussing this here is working. I'll setup a meeting on Monday
